File: solutions/go/lasagna-master/1/lasagna_master.go
package lasagna

func PreparationTime(layers []string, time_per_layer int) int {
	if time_per_layer == 0 {
		time_per_layer = 2
	}
	return len(layers) * time_per_layer
}

func Quantities(layers []string) (int, float64) {
	noodles := 0
	sauce := 0.0
	for _, layer := range layers {
		switch layer {
		case "noodles":
			noodles += 50
		case "sauce":
			sauce += 0.2
		}
	}
	return noodles, sauce
}

func AddSecretIngredient(friendsList, myList []string) {
	myList[len(myList)-1] = friendsList[len(friendsList)-1]
}

func ScaleRecipe(quantities []float64, portions int) []float64 {
	new_quantities := make([]float64, len(quantities))
	for i, q := range quantities {
		new_quantities[i] = q * float64(portions) / 2.0
	}
	return new_quantities
}

File: solutions/cpp/lasagna-master/1/lasagna_master.cpp
#include "lasagna_master.h"

#include <boost/range/adaptor/transformed.hpp>
#include <boost/range/algorithm.hpp>

using namespace boost::adaptors;
using boost::copy_range;
using namespace std;

namespace lasagna_master {

int preparationTime(vector<string> layers, int oneLayerTime) {
    return layers.size() * oneLayerTime;
}

amount quantities(vector<string> layers) {
    amount res{};
    for (const auto& layer : layers) {
        if (layer == "noodles") {
            res.noodles += 50;
        } else if (layer == "sauce") {
            res.sauce += 0.2;
        }
    }
    return res;
}

void addSecretIngredient(vector<string>& myList, vector<string> friendsList) {
    for (auto& ingredient : myList) {
        if (ingredient == "?") {
            ingredient = friendsList.back();
            return;
        }
    }
}

void addSecretIngredient(vector<string>& myList, string secret) {
    addSecretIngredient(myList, vector{secret});
}

vector<double> scaleRecipe(vector<double> input, int portions) {
    return copy_range<vector<double>>(input | transformed([&](double amount) {
                                          return amount * portions / 2;
                                      }));
}

}  // namespace lasagna_master
